experimental/pattern-matching.md

```diff
@@ -0,0 +1,119 @@
+# [Pattern Matching](proposal-pattern-matching)
+
+See also [AST examples](ast-examples) for this proposal.
+
+## MatchExpression
+
+```js
+interface MatchExpression <: Expression {
+    type: "MatchExpression";
+    discriminant: Expression;
+    id: Pattern | null;
+    clause: [ MatchClause ]
+}
+```
+
+## WhenClause
+```js
+interface MatchClause <: Node {
+    type: "MatchClause";
+    test: MatchPattern | null;
+    guard: Expression | null;
+    id: Pattern | null;
+    consequent: BlockStatement
+}
+```
+
+If `test` is `null`, `id` must be `null`.
+
+## MatchPattern
+```js
+interface MatchPattern <: Node {}
+```
+
+## ArrayMatchPattern
+```js
+interface ArrayMatchPattern <: MatchPattern {
+    type: "ArrayMatchPattern";
+    elements: [ MatchPattern ];
+}
+```
+
+## ObjectMatchPattern
+```js
+interface AssignmentMatchProperty <: Property {
+    type: "Property"; // inherited
+    value: MatchPattern;
+    kind: "init";
+    method: false;
+}
+
+interface ObjectMatchPattern <: MatchPattern {
+    type: "ObjectMatchPattern";
+    properties: [ AssignmentMatchProperty ];
+}
+```
+
+## RestMatchElement
+```js
+interface RestMatchElement <: MatchPattern {
+    type: "RestMatchElement";
+    argument: MatchPattern;
+}
+```
+
+## BinaryMatchPattern
+```js
+interface BinaryMatchPattern <: MatchPattern {
+    type: "BinaryPattern";
+    operator: "and" | "or" | "with";
+    left: MatchPattern;
+    right: MatchPattern;
+}
+```
+
+## AsMatchPattern
+```js
+interface AsMatchPattern <: MatchPattern {
+    type: "AsMatchPattern";
+    test: MatchPattern;
+    id: Pattern;
+}
+```
+
+## ExpressionMatchPattern
+```js
+interface ExpressionMatchPattern <: MatchPattern {
+    type: "ExpressionMatchPattern";
+    expression: Expression;
+}
+```
```

Comment:
not sure if this is covered or not; the syntax allows `^Identifier`, `^(Expression)`, but also things of the form `^A['B'].C()` or `^A?['B']?.C?.()`.

Reply:
does it matter if there are parentheses or not?

Reply:
Yes; any expression can be in parens, but only a specific limited set of forms can omit them.

Reply:
> the syntax allows ^Identifier, ^(Expression), but also things of the form ^A['B'].C() or ^A?['B']?.C?.().

Both `^Identifier` and `^(Expression)` are an ExpressionMatchPattern whose `expression` is an `Identifier` or an `Expression`. ESTree does not track whether parentheses are required to form a valid syntax. Instead a parser will throw when parentheses are required and a generator will add parentheses based on the AST structure.